scene/gui/flow_container.cpp

```diff
@@ -0,0 +1,249 @@
+			if (children_in_current_line > 0) {
+				ofs.x += separation_horizontal;
+			}
+			if (ofs.x + child_min_size.x > current_width) {
+				// Move in new line.
+				ofs.x = 0;
+				ofs.y += line_height + separation_vertical;
+				lines_data.push_back(Vector2i(children_in_current_line, line_height));
+
+				line_height = 0;
+				children_in_current_line = 0;
+			}
+
+			child_rects.push_back(Rect2(ofs, child_min_size));
+
+			if (child_min_size.y > line_height) {
+				line_height = child_min_size.y;
+			}
+
+			ofs.x += child_min_size.x;
+		}
+		children_in_current_line++;
+	}
+	lines_data.push_back(Vector2i(children_in_current_line, line_height));
+
+	// Second pass for in-line expansion and alignment.
+	int current_line_idx = 0;
+	int child_idx_in_line = 0;
+	for (int i = 0; i < get_child_count(); i++) {
+		Control *child = Object::cast_to<Control>(get_child(i));
+		if (!child || !child->is_visible_in_tree()) {
+			continue;
+		}
+		if (child->is_set_as_top_level()) {
+			continue;
+		}
+
+		Vector2i line_data = lines_data[current_line_idx];
+		if (child_idx_in_line >= lines_data[current_line_idx].x) {
+			current_line_idx++;
+			child_idx_in_line = 0;
+			line_data = lines_data[current_line_idx];
+		}
+		Rect2 child_rect = child_rects[i];
+
+		if (vertical) { /* VERTICAL */
+			if (child->get_h_size_flags() & SIZE_EXPAND) {
+				child_rect.size.width = line_data.height;
+			}
+			switch (line_alignment) {
+				case LINE_ALIGNMENT_TOP:
+					break;
+				case LINE_ALIGNMENT_CENTER:
+					child_rect.position.x += line_data.height / 2 - child_rect.size.width / 2;
+					break;
+				case LINE_ALIGNMENT_BOTTOM:
+					child_rect.position.x += line_data.height - child_rect.size.width;
+					break;
+			}
+		} else { /* HORIZONTAL */
+			if (child->get_v_size_flags() & SIZE_EXPAND) {
+				child_rect.size.height = line_data.height;
+			}
+			switch (line_alignment) {
+				case LINE_ALIGNMENT_TOP:
+					break;
+				case LINE_ALIGNMENT_CENTER:
+					child_rect.position.y += line_data.height / 2 - child_rect.size.height / 2;
+					break;
+				case LINE_ALIGNMENT_BOTTOM:
+					child_rect.position.y += line_data.height - child_rect.size.height;
+					break;
+			}
+		}
+		if (rtl) {
+			child_rect.position.x = get_rect().size.x - child_rect.position.x - child_rect.size.width;
+		}
+		fit_child_in_rect(child, child_rect);
+		child_idx_in_line++;
+	}
+	cached_height = (vertical ? ofs.x : ofs.y) + line_height;
+	cached_line_count = current_line_idx;
+}
+
+Size2 FlowContainer::get_minimum_size() const {
+	Size2i minimum;
+
+	for (int i = 0; i < get_child_count(); i++) {
+		Control *c = Object::cast_to<Control>(get_child(i));
+		if (!c) {
+			continue;
+		}
+		if (c->is_set_as_top_level()) {
+			continue;
+		}
+
+		if (!c->is_visible()) {
+			continue;
+		}
+
+		Size2i size = c->get_combined_minimum_size();
+
+		if (vertical) {
+			if (size.height > minimum.height) {
+				minimum.height = size.height;
+			}
+
+			minimum.width = cached_height;
```

Comment:
Seems like `cached_height` is used both for height and width. It should be renamed for better clarity, e.g. to `cached_size` or `cached_width`.

Reply:
You're right, the naming is suboptimal, but he naming is quite tricky here. 
I tried to find a better name for these variables, but since the variables are used for both the vertical and horizontal versions `cached_width` would have the same problems as `cached_height`. Although `cached_size` is less specific I think we could go with that.

Reply:
Ah right, I thought about "width" as "thickness", but that's actually wrong usage here. Yeah, size is ok.